### PR TITLE
BitFinex Feedhandler

### LINKED ISSUE #36: Writing a feedhandler
Write a feedhandler based on the the BitFinex cryptoexchange WebSocket API - https://bitfinex.readme.io/v2/docs/ws-general

A simple example of connecting to this API & receiving data can be seen here - https://gist.github.com/jonathonmcmurray/e8a3cd3b5b16e718daa2c26696d50d14

- [x]  a) You should provide functions to subscribe to trades and quotes for a given sym - should be possible to subscribe to one, two or all in one call
- [x]  b) There should be a utility function to retrieve a list of available syms (this may necessitate using REST API as well as WebSocket)
- [x]  c) Make use of appropriate channels based on subscription type - https://bitfinex.readme.io/v2/docs/ws-public
- [x]  d) Should be able to load a list of syms to subscribe to from a config file
- [x]  e) FH memory usage should be kept to a minimum
- [x]  f) Full TorQ integration (connection to TP via discovery, use TorQ logging where applicable etc.)
- [ ]  g) Unit tests
- [x]  h) Documentation

Extra points
- [ ] Batch mode
- [x] Timestamp data on entry

https://bitfinex.readme.io/v2/docs/ws-general
https://bitfinex.readme.io/v2/docs/ws-public

### PR DESCRIPTION
**Pull request by lholmes8**
## Description
Add a feedhandler for the BitFinex cryptoexchange API.
This pull request closes #36 .

## Checklist
Before creating a pull request please make sure you have the following:

- [x] A short but descriptive title
- [x] A description that references the issue e.g. this pull request closes (#issuenum).
- [x] Assignees responsible for moving the PR forward.
- [x] Colour-coded labels for categorisation.
- [x] A project to track PRs on a higher level

### CODE REVIEW

File: code/processes/bitfeed.q [hunk 1]
```diff
@@ -0,0 +1,58 @@
+us:@[value;`us;"General:
+    This script is to be used with TorQ as a feedhandler for BitFinex cryptoexchange API"]
+
+params:.Q.opt .z.x;
```
Comment (jonathonmcmurray) [inline code/processes/bitfeed.q]: shouldn't need this, TorQ does this already and puts in `.proc.params`

File: code/processes/bitfeed.q [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+if[`help in key params;-1 us;; exit 0];                                                                / Usage displays if -help argument is used when calling script
+
+.bfx.upd:show;
```
Comment (jonathonmcmurray) [inline code/processes/bitfeed.q]: don't think this should be here

Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Is this correct?

File: code/processes/bitfeed.q [hunk 3]
```diff
@@ -0,0 +1,58 @@
+us:@[value;`us;"General:
+    This script is to be used with TorQ as a feedhandler for BitFinex cryptoexchange API"]
+
+params:.Q.opt .z.x;
+if[`help in key params;-1 us;; exit 0];                                                                / Usage displays if -help argument is used when calling script
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Extra semi colon.

File: code/processes/bitfeed.q [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+.bfx.upd:show;
+
+.bfx.tickerplanttypes:@[value;`tickerplanttypes;`tickerplant];                                         / List of tickerplant types to try and make a connection to
+.bfx.replaylog:@[value;`replaylog;1b];                                                                 / Replay the tickerplant log file
+.bfx.schema:@[value;`schema;1b];                                                                       / Retrieve the schema from the tickerplant
+
+.bfx.tpconnsleepintv:@[value;`tpconnsleepintv;10];                                                     / Number of seconds between attempts to connect to the tp
+
+.bfx.notpconnected:{not `tickerplant in exec proctype from .servers.SERVERS where .dotz.liveh[w]}
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Remove squares.

File: code/processes/bitfeed.q [hunk 5]
```diff
@@ -0,0 +1,58 @@
+us:@[value;`us;"General:
+    This script is to be used with TorQ as a feedhandler for BitFinex cryptoexchange API"]
+
+params:.Q.opt .z.x;
+if[`help in key params;-1 us;; exit 0];                                                                / Usage displays if -help argument is used when calling script
+
+.bfx.upd:show;
+
+.bfx.tickerplanttypes:@[value;`tickerplanttypes;`tickerplant];                                         / List of tickerplant types to try and make a connection to
```
Comment (jonathonmcmurray) [inline code/processes/bitfeed.q]: I think the second arg in this and next two lines needs to have `.bfx.` at start, or else will be looking for these variables in root namespace

File: code/processes/bitfeed.q [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+.bfx.replaylog:@[value;`replaylog;1b];                                                                 / Replay the tickerplant log file
+.bfx.schema:@[value;`schema;1b];                                                                       / Retrieve the schema from the tickerplant
+
```
Comment (jonathonmcmurray) [inline code/processes/bitfeed.q]: why blank line here, looks like next line should be grouped with above

File: code/processes/bitfeed.q [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+.bfx.channels:()!();                                                                                   / Create empty dictionary to map syms to channel IDs
+.bfx.quotechannels:()!();
+
+.bfx.convertepoch:{`timestamp$(x*1000000)+1970.01.01D}
+.bfx.tradeupd:{
+  $[99h~type a:.j.k x;                                                                                 / If incoming data is in dictionary format
+    .bfx.channels[a`chanId]:a`pair;                                                                    / Insert channel ID and pair symbol into .bfx.channel
+    if["tu"~a 1;
+      a[0]:`$.bfx.channels a 0;                                                                        / Change channel ID to sym from .bfx.channels
+      .bfx.publish[`trade_bfx] enlist @[raze[a] 4 0 3 5 6;0;.bfx.convertepoch]]]};                      / Add sym, tradeID, size and price from data into .bfx.data
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Misaligned comment + put closing brackets on new lines.

File: code/processes/bitfeed.q [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+.servers.CONNECTIONS:.bfx.tickerplanttypes;
+
+while[.bfx.notpconnected[];
+    .os.sleep .bfx.tpconnsleepintv;
+    .servers.startup[];
+    .lg.o[`init;"searching for servers"];                                                              / Run the servers startup code again (to make connection to discovery)
+   ];
+
+.bfx.syms:@[value;`.bfx.syms;`];
+.bfx.allsyms:`$.j.k .Q.hg `:https://api.bitfinex.com/v1/symbols;	                                     / Calls for all available symbols from exchange
```
Comment (jonathonmcmurray) [inline code/processes/bitfeed.q]: I think spec requested a *function* to retrieve these on-demand, rather than them being pulled automatically

**Comment by claire-sloan** [comment]
change #47 to #issueNum or something, its linking to another issue in checklist

File: code/processes/bitfeed.q [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    if["tu"~a 1;
+      a[0]:`$.bfx.channels a 0;                                                                        / Change channel ID to sym from .bfx.channels
+      .bfx.publish[`trade_bfx] enlist @[raze[a] 4 0 3 5 6;0;.bfx.convertepoch]]]};                      / Add sym, tradeID, size and price from data into .bfx.data
+
+.bfx.quoteupd:{
+  $[99h~type a:.j.k x;                                                                                 / If incoming data is in dictionary format
+    .bfx.quotechannels[a`chanId]:a`pair;                                                               / Insert channel ID and pair symbol into .bfx.channel
+    if[not 10= type a 1;
+      a[0]:`$.bfx.quotechannels a 0;                                                                   / Change channel ID to sym from .bfx.channels
+      .bfx.publish[`quote_bfx] enlist raze[a] 0 1 2 3 4]]};                                            / Add sym, tradeID, size and price from data into .bfx.data
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Similar comments to above. Can the logic here be combined with `.bfx.tradeupd`? Seems a bit similar.

File: code/processes/bitfeed.q [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    .servers.startup[];
+    .lg.o[`init;"searching for servers"];                                                              / Run the servers startup code again (to make connection to discovery)
+   ];
+
+.bfx.syms:@[value;`.bfx.syms;`];
+.bfx.allsyms:`$.j.k .Q.hg `:https://api.bitfinex.com/v1/symbols;	                                     / Calls for all available symbols from exchange
+
+.bfx.tp:.servers.gethandlebytype[.bfx.tickerplanttypes;`any];
+
+.bfx.publish:{[t;x] .bfx.tp(".u.upd";t;flip x)};                                                       / Publish to tickerplant after replaylog
```
Comment (jonathonmcmurray) [inline code/processes/bitfeed.q]: Think it's neater to use `` `.u.upd`` rather than `".u.upd"`

File: database.q [hunk 11]
```diff
@@ -6,3 +6,5 @@ srcquote:([]time:`timestamp$(); sym:`g#`symbol$(); src:`symbol$(); bid:`float$()
 clienttrade:([]time:`timestamp$(); sym:`g#`symbol$(); price:`float$(); size:`int$();stop:`boolean$(); cond:`char$(); ex:`char$();side:`symbol$())
 pnl:([]time:`timestamp$();sym:`g#`symbol$();dcost:`float$();position:`long$();bid:`float$();ask:`float$();bidid:`long$();askid:`long$();pnl:`float$())
 bbook:([]time:`timestamp$(); sym:`g#`symbol$();bbid:`float$();bask:`float$();src_bask:`$();src_bbid:`$();bbid_id:`long$();bask_id:`long$();bsize:`long$();asize:`long$();bask_time:`timestamp$();bbid_time:`timestamp$())
+trade_bfx:([] time:`timestamp$();sym:`g#`symbol$();id:`float$();size:`float$();price:`float$())
+quote_bfx:([] time:`timestamp$();sym:`g#`symbol$();bbid:`float$();bsize:`float$();bask:`float$();asize:`float$())
```
Comment (ThomasSmyth) [inline database.q]: Better to avoid underscores in filenames if possible.

File: database.q [hunk 12]
```diff
@@ -6,3 +6,5 @@ srcquote:([]time:`timestamp$(); sym:`g#`symbol$(); src:`symbol$(); bid:`float$()
 clienttrade:([]time:`timestamp$(); sym:`g#`symbol$(); price:`float$(); size:`int$();stop:`boolean$(); cond:`char$(); ex:`char$();side:`symbol$())
 pnl:([]time:`timestamp$();sym:`g#`symbol$();dcost:`float$();position:`long$();bid:`float$();ask:`float$();bidid:`long$();askid:`long$();pnl:`float$())
 bbook:([]time:`timestamp$(); sym:`g#`symbol$();bbid:`float$();bask:`float$();src_bask:`$();src_bbid:`$();bbid_id:`long$();bask_id:`long$();bsize:`long$();asize:`long$();bask_time:`timestamp$();bbid_time:`timestamp$())
+trade_bfx:([] time:`timestamp$();sym:`g#`symbol$();id:`float$();size:`float$();price:`float$())
```
Comment (jonathonmcmurray) [inline database.q]: size should be a long, maybe id as well

File: database.q (hunk 11 above)
Comment (jonathonmcmurray) [inline database.q]: again, sizes should be longs, not floats

Comment (mkelly95) [reply database.q]: Sizes can be floats as they are currencies

Comment (jonathonmcmurray) [reply database.q]: fair point

File: code/processes/bitfeed.q [hunk 13]
```diff
@@ -0,0 +1,63 @@
+us:@[value;`us;"General:
+    This script is to be used with TorQ as a feedhandler for BitFinex cryptoexchange API"]
+
+if[`help in key .proc.params;-1 us; exit 0];                                                           / Usage displays if -help argument is used when calling script
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: whitespace

File: code/processes/bitfeed.q [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+.bfx.tickerplanttypes:@[value;`.bfx.tickerplanttypes;`tickerplant];                                    / List of tickerplant types to try and make a connection to
+.bfx.replaylog:@[value;`.bfx.replaylog;1b];                                                            / Replay the tickerplant log file
+.bfx.schema:@[value;`.bfx.schema;1b];                                                                  / Retrieve the schema from the tickerplant
+.bfx.tpconnsleepintv:@[value;`.bfx.tpconnsleepintv;10];                                                / Number of seconds between attempts to connect to the tp
+.bfx.notpconnected:{not `tickerplant in exec proctype from .servers.SERVERS where .dotz.liveh w}       / Check if tickerplant is connected and handle live
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: missing semi

File: code/processes/bitfeed.q [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    .servers.startup[];
+    .lg.o[`init;"searching for servers"];                                                              / Run the servers startup code again (to make connection to discovery)
+   ];
+
+.bfx.syms:@[value;`.bfx.syms;`];
+.bfx.getallsyms:{`$.j.k .Q.hg `:https://api.bitfinex.com/v1/symbols};                                  / Calls for all available symbols from exchange
+
+.bfx.tp:.servers.gethandlebytype[.bfx.tickerplanttypes;`any];
+
+.bfx.publish:{[t;x] .bfx.tp(`.u.upd;t;flip x)};                                                       / Publish to tickerplant after replaylog
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: whitespace + comment needs aligned

File: code/processes/bitfeed.q [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+.bfx.data:();                                                                                          / Create empty list to insert data into
+
+.bfx.channels:()!();                                                                                   / Create empty dictionary to map syms to channel IDs
+
+.bfx.quotechannels:()!();
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: add comment

File: code/processes/bitfeed.q [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+.bfx.convertepoch:{`timestamp$(x*1000000)+1970.01.01D}
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: semi colon + comment

File: code/processes/bitfeed.q [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+.bfx.tradeupd:{
+  $[99h~type a:.j.k x;                                                                                 / If incoming data is in dictionary format
+    .bfx.channels[a`chanId]:a`pair;                                                                    / Insert channel ID and pair symbol into .bfx.channel
+    if["tu"~a 1;
+      a[0]:`$.bfx.channels a 0;                                                                        / Change channel ID to sym from .bfx.channels
+      .bfx.publish[`trade_bfx] enlist @[raze[a] 4 0 3 5 6;0;.bfx.convertepoch]                         / Add sym, tradeID, size and price from data into .bfx.data
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: whitespace

File: code/processes/bitfeed.q (hunk 17 above)
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Can you reverse the arguments to the sum to remove the brackets?

File: code/processes/bitfeed.q [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+      .bfx.publish[`quote_bfx] enlist raze[a] 0 1 2 3 4                                                / Add sym, tradeID, size and price from data into .bfx.data
+      ]
+    ]
+  };
+
+.bfx.trade:.ws.open["wss://api.bitfinex.com/ws/2";`.bfx.tradeupd];
+.bfx.quote:.ws.open["wss://api.bitfinex.com/ws/2";`.bfx.quoteupd];
+
+.bfx.sub:{.bfx.subone[x;y] each $[z~`;.bfx.getallsyms[];z inter .bfx.getallsyms[]]}
+.bfx.unsub:{.bfx.unsubone[x;y] each $[z~`;.bfx.getallsyms[];z inter .bfx.getallsyms[]]}
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Can probably do something like:
```
$[z~`;;z inter].bfx.getallsyms[];
```

File: code/processes/bitfeed.q [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+      ]
+    ]
+  };
+
+.bfx.quoteupd:{
+  $[99h~type a:.j.k x;                                                                                 / If incoming data is in dictionary format
+    .bfx.quotechannels[a`chanId]:a`pair;                                                               / Insert channel ID and pair symbol into .bfx.channel
+    if[not 10= type a 1;
+      a[0]:`$.bfx.quotechannels a 0;                                                                   / Change channel ID to sym from .bfx.channels
+      .bfx.publish[`quote_bfx] enlist raze[a] 0 1 2 3 4                                                / Add sym, tradeID, size and price from data into .bfx.data
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: The layout of this section makes it easier to read, but I'm still now a huge fan :+1:

Tip: if you're modifying an index but immediately using the whole list you can use `@` apply instead, for example something like:
```
raze[@[a;0;`$.bfx.quotechannels]]0 1 2 3 4
```

File: code/processes/bitfeed.q [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+      ]
+    ]
+  };
+
+.bfx.trade:.ws.open["wss://api.bitfinex.com/ws/2";`.bfx.tradeupd];
+.bfx.quote:.ws.open["wss://api.bitfinex.com/ws/2";`.bfx.quoteupd];
```
Comment (jonathonmcmurray) [inline code/processes/bitfeed.q]: Shouldn't need to open two connections, you can get quotes & trades over the same one

**Comment by ThomasSmyth** [comment]
Needs to be "closes #36" in the description for automatic closure.

File: code/processes/bitfeed.q [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+  $[99h~type a:.j.k x;                                                                                 / If incoming data is in dictionary format
+    .bfx.quotechannels[a`chanId]:a`pair;                                                               / Insert channel ID and pair symbol into .bfx.channel
+    if[not 10=type a 1;
+      a[0]:`$.bfx.quotechannels a 0;                                                                   / Change channel ID to sym from .bfx.channels
+      .bfx.publish[`quote_bfx]enlist raze[a] 0 1 2 3 4                                                 / Add sym, tradeID, size and price from data into .bfx.data
+      ]
+    ]
+  };
+
+.bfx.sub:{.bfx.subone[x;y] each $[z~`;.bfx.getallsyms[];z inter .bfx.getallsyms[]]};                   / Subscribe to each sym
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: I made the comment on 55, but this is the same, can use the same conditional here too.

File: code/processes/bitfeed.q [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+.bfx.unsub:{.bfx.unsubone[x;y] each $[z~`;;z inter].bfx.getallsyms[]};                                 / Unsubscribe to each sym
+.bfx.subone:{@[{[x;y;z]x .j.j `event`channel`pair!(`subscribe;y;z)}[;y;z];x;{[x].bfx.resetHandles[]}   / Subscribe to individual sym and reset handle if necessary
```
Comment (ThomasSmyth) [inline code/processes/bitfeed.q]: Do you need explicitly declare the params `[x;y;z]` and `[x]`?

**Pull request merged**